src/Authentication/Wsse.php

```diff
@@ -22,25 +23,39 @@ final class Wsse implements Authentication
      */
     private $password;
 
+    /**
+     * @var string
+     */
+    private $hashAlgorithm;
+
+    /**
+     * @var array<string>
+     */
+    private $acceptedHashAlgos = ['sha1', 'sha512', 'sha384', 'sha3-384', 'sha3-512'];
+
     /**
      * @param string $username
      * @param string $password
+     * @param string $hashAlgorithm To use a better hashing algorithm than the weak sha1, pass the algorithm to use, e.g. "sha512"
      */
-    public function __construct($username, $password)
+    public function __construct($username, $password, $hashAlgorithm = 'sha1')
     {
         $this->username = $username;
         $this->password = $password;
+        $this->hashAlgorithm = $hashAlgorithm;
     }
 
     /**
      * {@inheritdoc}
      */
     public function authenticate(RequestInterface $request)
     {
-        // TODO: generate better nonce?
         $nonce = substr(md5(uniqid(uniqid().'_', true)), 0, 16);
         $created = date('c');
-        $digest = base64_encode(sha1(base64_decode($nonce).$created.$this->password, true));
+        if (false === in_array($this->hashAlgorithm, $this->acceptedHashAlgos)) {
+            throw new InvalidArgumentException(sprintf('Unaccepted hashing algorithm: %s', $this->hashAlgorithm));
```

Comment:
i would not do this. if the user specifies an algorithm that is not supported, there is a clear error message: https://3v4l.org/ju4q7

Reply:
This issue you specify I've resolved. I create the `acceptedHashAlgos` variable and check whether the hashing algorithm is accepted.

Reply:
why do you restrict the algorithms to that list? there is [hash_algos](https://www.php.net/manual/en/function.hash-algos.php) that would provide the full list of algorithms that is supported by the installation - but i would not even check the specified algorithm and instead let the `hash` method create the error if something invalid has been specified.

if you want to check with hash_algos, we should do it in the constructor to fail early, at the place where the invalid parameter is passed, rather than some point later.

Reply:
Using [hash_algos](https://www.php.net/manual/en/function.hash-algos.php) function will let this class support many hashing algorithms. Including `md5`, `sha1`, `sha2` and so on.

That's the reason why I use the array to restrict accepted hashing algorithms.

If you have the proper approach, please provide it. Thanks.

Reply:
Or we can leave hashing algorithms once the specific algorithm is valid for `hash` function.

Developers should take their own risks if using the weak hashing algorithms.

And we only do hashing algorithm checking whether it's valid on `hash_algos` during this class constructor.

@dbu, what do you think?

Reply:
i was trying to find a proper specification of which algorithms are allowed here, but did not find anything. does WSSE specify which algorithms are allowed?
the only thing i found is https://www.oasis-open.org/committees/download.php/13392/wss-v1.1-spec-pr-UsernameTokenProfile-01.htm and that talks about sha1 as default algorithm or unspecified "other" algorithms. what is the source for the list in acceptedHashAlgos in this PR? if there is some extension to WSSE specifying exactly these algorithms, then the list is of course right to do.

Reply:
if technically anything is allowed, i would leave it up to the user of the library. we can do a phpdoc in the constructor argument to list recommended algorithms, but in the end it should be up to the dev. and if you e.g. have to talk with a wsse backend that you have no control over, you might just have to work with a poor algorithm.

Reply:
Thanks. I use the `hash_algos` function to check whether specific hashing algorithm is valis.